activerecord/test/cases/serialized_attribute_test.rb

```diff
@@ -392,6 +392,68 @@ def test_nil_is_always_persisted_as_null
     assert_equal [topic], Topic.where(content: nil)
   end
 
+  class EncryptedType < ActiveRecord::Type::Text
+    include ActiveModel::Type::Helpers::Mutable
+
+    attr_reader :subtype, :encryptor
+
+    def initialize(subtype: ActiveModel::Type::String.new)
+      super()
+
+      @subtype   = subtype
+      @encryptor = ActiveSupport::MessageEncryptor.new("abcd" * 8)
+    end
+
+    def serialize(value)
+      subtype.serialize(value).yield_self do |cleartext|
+        encryptor.encrypt_and_sign(cleartext) unless cleartext.nil?
+      end
+    end
+
+    def deserialize(ciphertext)
+      encryptor.decrypt_and_verify(ciphertext)
+        .yield_self { |cleartext| subtype.deserialize(cleartext) } unless ciphertext.nil?
+    end
+
+    def changed_in_place?(old, new)
+      if old.nil?
+        !new.nil?
+      else
+        deserialize(old) != new
+      end
+    end
+  end
+
+  def test_decorated_type_with_type_for_attribute
+    old_registry = ActiveRecord::Type.registry
+    ActiveRecord::Type.registry = ActiveRecord::Type.registry.dup
+    ActiveRecord::Type.register :encrypted, EncryptedType
+
+    klass = Class.new(ActiveRecord::Base) do
+      self.table_name = Topic.table_name
+      store :content
+      attribute :content, :encrypted, subtype: type_for_attribute(:content)
```

Comment:
Do users have to provide a subtype or can it be inferred automatically? Feels like attribute should automatically cascade the types.

Reply:
Do you mean that we should provide a way to inject a subtype to the type which has `subtype` option?

https://github.com/rails/rails/blob/70259f5e519869052a25db2c832e2c9d53349c66/activerecord/test/cases/serialized_attribute_test.rb#L395-L400

https://github.com/rails/rails/blob/70259f5e519869052a25db2c832e2c9d53349c66/activerecord/test/cases/serialized_attribute_test.rb#L446-L450

Reply:
I'm not sure what you mean by "automatically cascade the types".

In 6.1, there is no reliable way. If an attribute isn't decorated (no `enum`, `serialize`, `store` are used), `attribute(:content) { |subtype| EncryptedType.new(subtype: subtype) }`, and `decorate_attribute_type("content") { |subtype| EncryptedType.new(subtype: subtype) }` works.

In main branch, `attribute(:content) { |subtype| EncryptedType.new(subtype: subtype) }` works even if `enum`, `serialize`, `store` are used on the attribute (`decorate_attribute_type` is removed in the branch).

If we provide a way for decorator officially, I feel that it is better to provide the API separately (e.g. `decorate_attribute` etc) from the `attribute`.

Reply:
Wups, sorry, I forgot to get back to this! I'd just like to be able to do:

```ruby
store :content
attribute :content, :encrypted
```

without needing to pass the `subtype`, `attribute :content, :encrypted, subtype: type_for_attribute(:content)`.

> If we provide a way for decorator officially, I feel that it is better to provide the API separately (e.g. `decorate_attribute` etc) from the `attribute`.

That also sounds good 👍